flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateHandleStreamImpl.java

```diff
@@ -44,23 +46,42 @@
 
     private final long size;
     private final long incrementalSize;
+    private final StateHandleID stateHandleID;
 
     public ChangelogStateHandleStreamImpl(
             List<Tuple2<StreamStateHandle, Long>> handlesAndOffsets,
             KeyGroupRange keyGroupRange,
-            long size) {
-        this(handlesAndOffsets, keyGroupRange, size, size);
+            long size,
+            long incrementalSize) {
+        this(
+                handlesAndOffsets,
+                keyGroupRange,
+                size,
+                incrementalSize,
+                new StateHandleID(UUID.randomUUID().toString()));
```

Comment:
Creating a random ID here doesn't seem very straightforward to me, when looking at the call sites.

Especially, `MetadataV2V3SerializerBase` - why doesn't it store this ID? I guess it's fine because currently it's only used inside this handle itself; but if ID will be used for other purposes then it may become an issue.

So I'd propose to have an explicit contructor argument (at least for production code); and store this ID in the metadata.

WDYT?

Reply:
I have already introduced a static `restore` method to include the stateHandleId, I am not sure whether this is enough here. And what do you mean the `metadata`, do you mean persisting the state handle id on meta data seraizlier serializing? Actually, I have already persisting this on serializing.

Reply:
Sorry, misplaced this comment.
[Moved](https://github.com/apache/flink/pull/18391/files#r803666621) to `ChangelogStateBackendHandleImpl`.